### DISCUSSION TITLE
PrismaClient in Express with many files

### DISCUSSION THREAD

**Discussion by florianess** [prisma/prisma#3700] [Q&A] [answered]
Hi team, I'm new to Prisma and I want to learn the best practices

### Question

I'm building a Express Typescript API with separate files for routes, controllers, services. I don't know if I have to generate PrismaClient (`const prisma = new PrismaClient`) only once at the beginning of my app and then export in differente files. Or can I just create on the fly the PrismaClient in each services?

Do you have any recommendation / example ? Your TS express example is only in one file so pretty simple.

Thanks,
Have a great day

**Comment by ryands17** [comment, marked as answer]
@florianess 👋 
The best way would be to initialize `PrismaClient` only one in a separate file and import it wherever required. That's the recommended approach.

**Reply by SharadKumar** [reply]
Hi Ryan
I didnt find this anywhere explicitly stated in the docs. It would be good to have this guidance... a sample with how to best manage connections on a typical project. When multiple files using the client, how to $disconnect. Nuances (if any) with serverless.

**Reply by ryands17** [reply]
@SharadKumar Yes, currently this isn't mentioned in the docs. We can add that to the docs in a specific part :)

Also regarding managing connections especially with Serverless, we do not have any best practice for that yet, but there are a couple of things that you can try.

1. It's best to not directly connect to the DB and use something like [PgBouncer](https://www.pgbouncer.org/) to handle DB connections. Then you can try by initializing PrismaClient outside your Serverless handler so that it is re-used on subsequent requests.

2. Try setting the `connection_limit` parameter to a lower value and then run tests to see if that helps. You can read more about this paramter [here](https://www.prisma.io/docs/reference/database-connectors/postgresql/#connection-details).

For long-running servers, just a single instance of PrismaClient initialized in your main file and used everywhere is totally fine.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
